rust/arrow/src/array/value_iterator.rs

```diff
@@ -0,0 +1,305 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::{
+    Array, ArrayRef, BinaryOffsetSizeTrait, BooleanArray, GenericBinaryArray,
+    GenericListArray, GenericStringArray, OffsetSizeTrait, StringOffsetSizeTrait,
+};
+
+/// an iterator that returns Some(bool) or None.
+// Note: This implementation is based on std's [Vec]s' [IntoIter].
+#[derive(Debug)]
+pub struct BooleanValueIter<'a> {
```

Comment:
For all except PrimitiveArray, these effectively duplicate the existing 'iterator of options' array iterators, except they provide the raw value without validity information of any kind.  
It is to allow access for kernels that prefer to access all values (ex: if an operation is fast AND infallible, it can avoid accessing the null array just to branch so it can ignore it).  I am not sure if this need comes up very often outside of the primitive or boolean array.